### PR TITLE
Various cleanup: warning fixes, function names, better debugging commands

### PR DESCRIPTION
**Pull request by davidsoncolin**
Types
Changed int to size_t and size_t to int to remove warnings.
Explicitly cast float to int to remove warnings.
Explicitly case size() to int to remove warnings.

Clearer function names
Changed UCI::move and UCI::square to UCI::to_string(Move) and UCI::to_string(Square).
Changed Position::set to Position::init.
Removed Utils::lcm
Removed using namespace std.

Enhancement
Added more info to board ASCII printout (move number, turn and declare check)
Added debugging commands "eval" and "testgame".
Improvements to debugging commands.
Debug command 'play' now accepts a sequence of moves.
Debug command 'showgame' now emits a string that recreates the board state, including the last 8 moves.
Checkmate/Stalemate/Draw/Check now properly declared when moving with 'go' command.
Added Position::has_no_moves.

Bugfixes
Crash when 'go' command issued at end of game
Missing underscore WIN32

Project settings
Moved the shared defines to the left.

### CODE REVIEW

File: src/UCI.cpp [hunk 1]
```diff
@@ -309,65 +335,82 @@ void UCI::loop(const std::string& start) {
...
+      else if (token == "testgo") {
+          // compare the result with "no history"
+          NNCache::get_NNCache().resize(0);
+          BoardHistory b;
+          b.positions.push_back(bh.cur());
+          go(*search, b, is);
+          bh.do_move(b.cur().get_move());
+          NNCache::get_NNCache().resize(0);
+      }
+      else if (token == "reset") {
```
Comment (killerducky) [inline src/UCI.cpp]: I think this should be what the ucinewgame command does?

File: src/UCTSearch.cpp [hunk 2]
```diff
@@ -325,6 +324,7 @@ Move UCTSearch::think(BoardHistory&& new_bh) {
         dump_analysis(milliseconds_elapsed, true);
     }
     Move bestmove = get_best_move();
+    myprintf("selected %s [%s]\n", UCI::to_string(bestmove).c_str(), bh_.cur().move_to_san(bestmove).c_str());
```
Comment (killerducky) [inline src/UCTSearch.cpp]: I think this is redundant with UCI.cpp which also outputs bestmove? I see you included move_to_san, if that's the goal maybe move this to UCI.cpp instead?

**Comment by killerducky** [comment]
I saw you added a "reset" command, but I think that's the same as "ucinewgame" command, so I think you should move it there. Here's the reference I've been using:
http://wbec-ridderkerk.nl/html/UCIProtocol.html There are a few other things we do that are not fully UCI complaint, hope we can work on those soon.

Also now with treereuse being merged if you really want to reset things you need to clear the search tree in addition to clearing NNCache. The easy way is to just construct a new search object. Or maybe we should have a method in search that does it, and maybe that method is responsible for clearing NNCache? Not sure what is best.

Also this is a massive change in terms of lines touched. It would be easier to review if you split it into 2 commits. One of mostly non-functional name and white-space changes. And different one with functional changes like adding commands. This will ensure reviewers don't scroll past a bunch of trivial changes and not notice there is a functional change in the middle of it.

File: src/UCI.cpp [hunk 3]
```diff
@@ -309,65 +335,82 @@ void UCI::loop(const std::string& start) {
...
-	  }
-		  //else if (token == "eval")  sync_cout << Eval::trace(pos) << sync_endl;
-		  else if (token != "quit") {
-		  myprintf_so("Unknown command: %s\n", cmd.c_str());
-	  }
+          std::stringstream ss;
+          ss << bh.cur();
+          myprintf("%s\n", ss.str().c_str());
+      }
+      else if (token == "testgo") {
```
Comment (killerducky) [inline src/UCI.cpp]: Maybe users can do "ucinewgame" and "go"? Then we don't need a separate command. Or is there another reason for this command?

**Comment by davidsoncolin** [comment]
I'm not certain what "ucinewgame" is meant to do, but if it's exactly what "reset" does then by all means make it a synonym.
"testgo" is for comparing the result of running the net without the history of the previous 7 games (without destroying that history). I was looking for positions where the history makes a difference.
all the functional changes should be in the file UCI.cpp, apart from the bugfix of the missing underscore in main.cpp.

**Comment by killerducky** [comment]
Oh I see your comment in testgo says that already, nice feature! Maybe we can find a better name? An argument to the go command perhaps `go nohistory`?

Looks good to me.

**Comment by killerducky** [comment]
Hey sorry we never got around to merging this. UCI.cpp has been reworked extensively so much of this is not relevant anymore. Feel free to make another PR.

**Pull request closed without merging**